v1.9 (focused backport) backports 2021-06-28

aditighag opened the pull request:
* #16259 -- kpr: Enable/fix Cilium socket based load balancing in different environments (@aditighag)
* #16815 -- Fixes for platforms where init container fails to run -- Fixes for platforms where init container fails to run (@aditighag)

Once this PR is merged, you can update the PR labels via:
```upstream-prs
$ for pr in 16259 16815; do contrib/backporting/set-labels.py $pr done 1.9; done
```

aditighag committed:
cilium-daemonset: Fix ineffective socket-lb caused by incorrect cgroup2 fs mount

[ upstream commit fa8bea45562f7ea3005708e968c419720a0ad190 ]

If container runtimes are run with cgroup v2, Cilium agent pod would be deployed in a
separate cgroup namespace. For example, Docker container runtime with cgroupv2 support
switched to private cgroup namespace mode as the default [1]. Due to cgroup namespaces [2],
the cgroup fs mounted by the Cilium pod points to a virtualized cgroup hierarchy instead of
the host cgroup root. As a result, BPF programs are attached to the nested cgroup root, and socket-lb
isn't effective for other pods.

Fix:

Mount cgroup2 fs from the host so that BPF programs are attached at the host cgroup root.
A new init container is added to the Cilium Daemonset that mounts cgroup2 fs on the host.
The `/proc/1/ns/` directory on the host is required to be mounted so that cgroup and mount
namespaces are enabled as enterable namespaces while running the `nsenter` command.
Additionally, cgroup2 fs can be attached to different paths so let's mount it on the host
at a cilium-specific custom location. Cilium can thus have control over the location
(e.g., create the directory if it doesn't exist). This also helps in effectively identifying
if a cgroup2 mount already exists at the custom location.

[1] https://docs.docker.com/config/containers/runmetrics/#running-docker-on-cgroup-v2
[2] https://man7.org/linux/man-pages/man7/cgroup_namespaces.7.html

Reported-By: Kornilios Kourtis <kornilios@isovalent.com>
Fixes: #15137

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
Revert "cgroups: Determine cgroup v2 hierarchy root for Kind"

[ upstream commit 0c166f6c6488d3ef12afb6015a86cdd222c890e9 ]

This reverts commit e9ce8306400bf416087046b8d5b013b23ebdcb3e.

This logic is no longer needed as we mount cgroup v2 filesystem
from the underlying kubernetes node. This will enable cilium to
correctly attach BPF programs at every `kind` node's cgroup root.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
docs: Document failure scenario for kind deployment

[ upstream commit ca8456ca5606bc03d643bd4eaccaab751b742f06 ]

Deploying a kind cluster in an environment where Cilium is already
running (for example, in the Cilium development VM) can lead to Cilium
pods crashing. This can also happen if there are other BPF cgroup programs
attached to the parent ``cgroup`` hierarchy of the kind container nodes.

Relevant Linux kernel code reference - https://elixir.bootlin.com/linux/latest/source/kernel/bpf/cgroup.c#L457.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
docs: Add troubleshooting steps to the kube-proxy free guide

[ upstream commit f263235b4e3f0de9dbddbf6353f65dbd6c0ad036 ]

Document the requirement that Cilium agent needs to be able
to attach BPF cgroup programs at the host cgroup root, in order
for socket-based load balancing (aka host-reachable services)
to be effective for other pods and host processes.

More details in the PR - https://github.com/cilium/cilium/pull/16259

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
defaults: Update default cgroup root

[ upstream commit 8b9bc2ed952533d296066ed711ddfae06c2c7ed4 ]

`/var/run` is a symlink to `/run` on most platforms, and may
not always be present. Also, this is consistent with the
`DefaultMapRootFallback` currently configured in the agent.

Example -

$ sudo mount -t cgroup2 none /var/run/cilium/cgroupv2
$ mount | grep cgroup
none on /run/cilium/cgroupv2 type cgroup2 (rw,relatime)

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
cilium-daemonset: Host cgroup root mount as alternative to auto-mount

[ upstream commit 826531447fa5ba18d2fe0df8d7eed9881b47235d ]

Cilium agent daemonset auto mounts cgroup2 filesystem on the host by default.
However, it needs to mount host's `/proc` inside an init container
in order to do that. To disable this auto-mount behavior, we introduce a helm
option. When auto-mount is disabled, users can specify the mount point on the
underlying host where cgroup v2 fs is already mounted. We then volume mount this
directory inside the cilium agent pod.
The reason why we don't set the host cgroup2 mount point to a hard-coded path
such as `/sys/fs/cgroup`, is because cgroup2 filesystem mount point can be platform
dependent. See this note in the cgroup manpage [1] -

>Note that on many modern systems, systemd(1) automatically mounts
the cgroup2 filesystem at /sys/fs/cgroup/unified during the boot
process.

[1] https://man7.org/linux/man-pages/man7/cgroups.7.html

Suggested-by: Kornilios Kourtis <kornilios@isovalent.com>.
Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag commented on Documentation/gettingstarted/kubeproxy-free.rst:
```diff
@@ -236,16 +251,16 @@ port ``31940`` (one for each of devices ``eth0`` and ``eth1``):
 .. parsed-literal::
 
     kubectl exec -it -n kube-system cilium-fmh8d -- cilium service list
-    ID   Frontend               Service Type   Backend                    
```
I have no idea why these changes are showing up. When I do `git annotate`, I don't see them.

aanm replied:
I think it's because your editor is removing the trailing spaces. It's fine.

aditighag committed:
install/kubernetes: Mount host's `/proc` in init container

[ upstream commit eecca2e0d912b5c17a998c317b4c3d3e4c142e89 ]

The mount-cgroup init container needs to run a mount command
on the underlying host. But the current approach to mount
`/proc/1/ns` fails on distros like Fedora when running on
 minikube -

mounting "/proc/1/ns" to rootfs at "/var/lib/docker/.../merged/hostpid1ns" caused: permission denied: unknown

Mount host's `/proc` instead.

Fixes: fa8bea45562f ("cilium-daemonset: Fix ineffective socket-lb caused by incorrect cgroup2 fs mount")

Reported-By: André Martins <andre@cilium.io>
Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
install/kubernetes: Set image pull policy for init container

[ upstream commit f9b79c7bb776653fdc002eb8bfbff051d36156ad ]

Fixes: fa8bea45562f ("cilium-daemonset: Fix ineffective socket-lb caused by incorrect cgroup2 fs mount")

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
install/kubernetes: Remove `sh` and `mount` dependency from init container

[ upstream commit a76bbde4320591399cd94392ab39650737ee4e13 ]

The mount-cgroup init container runs a mount command on the underlying
host using `nsenter`. However, certain distros like Talos don't have `sh`
or `mount` utilities available. Hence, move the logic to check and mount
cgroup2 fs to a statically linked Go program binary.

Fixes: fa8bea45562f ("cilium-daemonset: Fix ineffective socket-lb caused by incorrect cgroup2 fs mount")

Signed-off-by: Aditi Ghag <aditi@cilium.io>
